{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ce0a399e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def loadFindThomsRes(pth, refSketch):\n",
    "    compResPerRead = {}\n",
    "    \n",
    "    for l in open(pth, 'r'):\n",
    "        l = l.strip()\n",
    "\n",
    "        if l.startswith('s'):\n",
    "            lastRdId = int(l.split('_')[1])\n",
    "            compResPerRead[lastRdId] = []\n",
    "        else:\n",
    "            cols = l.split(' ')\n",
    "            s = refSketch[int(cols[1])][0] + 1 - K\n",
    "            e = refSketch[int(cols[3])][0]\n",
    "            compResPerRead[lastRdId].append({\"id\": lastRdId, \"start\": s, \"end\": e, \"score\": float(cols[5])})\n",
    "        \n",
    "    return compResPerRead"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "a6cba583",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import deque\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "NT_IN_BITS = {'A': 0, 'C': 1, 'G': 2, 'T': 3}\n",
    "\n",
    "#This function calculates the minimizer sketch of a sequence. It is influenced by the code of \"The minimizer Jaccard\n",
    "#estimator is biased and inconsistent.\" from Belbasi et al. (function \"winnowed_minimizers_linear(perm,windowSize)\" \n",
    "#in file \"winnowed_minimizers.py\").\n",
    "def calcMiniSketch(seq, k, w):\n",
    "    sketch = []\n",
    "    #A deque to store k-mer hashes inside the current window\n",
    "    windowKmers = deque()\n",
    "    mask = (4 ** k) - 1\n",
    "    lastIdx = -1\n",
    "\n",
    "    for i in range(len(seq) - k + 1):\n",
    "        kmerBits = 0\n",
    "        kmerBitsRevComp = 0\n",
    "        windowBorder = i - (w - 1)\n",
    "        \n",
    "        #Get bit representation of k-mer\n",
    "        for c in seq[i:i+k]:\n",
    "            kmerBits = (kmerBits << 2) + NT_IN_BITS[c]\n",
    "\n",
    "        #Get bit representation of k-mer's reverse complement\n",
    "        for c in str(Seq(seq[i:i+k]).reverse_complement()):\n",
    "            kmerBitsRevComp = (kmerBitsRevComp << 2) + NT_IN_BITS[c]\n",
    "\n",
    "        #If a k-mer is its own reverse complement we skip it\n",
    "        if kmerBits == kmerBitsRevComp:\n",
    "            continue\n",
    "\n",
    "        #Depending on which hash is smaller we consider either a k-mer or its reverse complement per position\n",
    "        if kmerBits < kmerBitsRevComp:\n",
    "            kmer = (i, kmerBits, getHash(kmerBits, mask))\n",
    "        else:\n",
    "            #A k-mer is a pair of k-mer's start position and its hash\n",
    "            kmer = (i, kmerBitsRevComp, getHash(kmerBitsRevComp, mask))\n",
    "            \n",
    "        #Remove all k-mers with a hash value larger than the newly calculated one\n",
    "        while (len(windowKmers) > 0) and (windowKmers[-1][2] > kmer[2]):\n",
    "            windowKmers.pop()\n",
    "\n",
    "        #Save new k-mer as window k-mer\n",
    "        windowKmers.append(kmer)\n",
    "\n",
    "        #Remove k-mer if it is not any longer inside the window\n",
    "        while (len(windowKmers) > 0) and (windowKmers[0][0] < windowBorder):\n",
    "            windowKmers.popleft()\n",
    "\n",
    "        #As soon as we have seen a first full window of k-mers choose a minimizer\n",
    "        if (windowBorder >= 0) and (len(windowKmers) > 0):      \n",
    "            #We do not choose the same minimizer for a second time\n",
    "            if lastIdx != windowKmers[0][0]:\n",
    "                sketch.append((windowKmers[0][0]+k-1, windowKmers[0][1], windowKmers[0][2]))\n",
    "                lastIdx = windowKmers[0][0]\n",
    "                \n",
    "            while len(windowKmers) > 1 and windowKmers[0][1] == windowKmers[1][1]:\n",
    "                windowKmers.popleft()\n",
    "                sketch.append((windowKmers[0][0]+k-1, windowKmers[0][1], windowKmers[0][2]))    \n",
    "                lastIdx = windowKmers[0][0] \n",
    "\n",
    "    #If our sequence was too small to get a full window of k-mers to consider take the smallest one found so far\n",
    "    if windowBorder < 0 and len(windowKmers) > 0:\n",
    "        sketch.append((windowKmers[0][0]+k-1, windowKmers[0][1], windowKmers[0][2]))\n",
    "        \n",
    "        while len(windowKmers) > 1 and windowKmers[0][1] == windowKmers[1][1]:\n",
    "            windowKmers.popleft()\n",
    "            sketch.append((windowKmers[0][0]+k-1, windowKmers[0][1], windowKmers[0][2]))\n",
    "\n",
    "    return sketch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3b2af129",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function filters out best fitting intervals from a given set of homology intervals. Given a set of nested\n",
    "#homology intervals, an interval is \"best fitting\" if its length is closest to the length of the corresponding read\n",
    "#Parameter: intCoords: A dictionary with integer-based read ids as keys and lists of pairs representing start and\n",
    "#                      end coordinates of intervals as values\n",
    "def filterBestFittingInts(intCoords):\n",
    "    bestFitting = {}\n",
    "    \n",
    "    #For each read...\n",
    "    for r in intCoords:\n",
    "        bestFitting[r] = []\n",
    "        #Clusters of nested intervals are stored as a dictionary with the coordinated of the outermost interval as \n",
    "        #keys and a list of all interval pairs as values\n",
    "        nestedInts = {}\n",
    "        \n",
    "        #...cluster nested intervals together\n",
    "        for i in intCoords[r]:\n",
    "            for c in nestedInts:\n",
    "                #Check if the current interval covers the outermost interval of this nested cluster\n",
    "                if i[0] <= c[0] and i[1] >= c[1]:\n",
    "                    nestedInts[i] = list(nestedInts[c]) + [i]\n",
    "                    del nestedInts[c]\n",
    "                    break\n",
    "                #Check if the current interval is covered by an outermost interval of some cluster already\n",
    "                elif i[0] >= c[0] and i[1] <= c[1]:\n",
    "                    nestedInts[c].append(i)\n",
    "                    break\n",
    "                \n",
    "            #Create a new cluster\n",
    "            nestedInts[i] = [i]\n",
    "            \n",
    "        #Select the best fitting interval of each cluster\n",
    "        readLength = len(rdSeqs[f\"s_{r}\"].seq)\n",
    "\n",
    "        for c in nestedInts:\n",
    "            minDiff = abs((c[1] - c[0]) - readLength)\n",
    "            bestInt = c\n",
    "                \n",
    "            for i in nestedInts[c]:\n",
    "                diff = abs((i[1] - i[0]) - readLength)\n",
    "                    \n",
    "                if diff < minDiff:\n",
    "                    minDiff = diff\n",
    "                    bestInt = i\n",
    "                    \n",
    "            bestFitting[r].append(bestInt)\n",
    "    \n",
    "    return bestFitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a6887341",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function returns a dictionary with integer-based read ids as keys and a list of start stop tupels of homology \n",
    "#intervals reported by FindThoms as values. Stop coordinates may be corrected in the way that the k-1 overlap of the\n",
    "#last k-mer part of the alpha-homology is not included anymore.\n",
    "#Parameter: \n",
    "#res: Result of FindThoms as returned from function loadFindThomsRes\n",
    "#corrEnd: Flag indicating if the end coordinate shall be adaptated\n",
    "def parseCoordsFromFindThomsRes(res, corrEnd):\n",
    "    ints = {}\n",
    "        \n",
    "    for r in res:\n",
    "        if corrEnd:\n",
    "            #Start position was corrected inside loadFindThomsRes already\n",
    "            ints[r] = [(h[\"start\"], h[\"end\"] + 1 - 15) for h in res[r]]\n",
    "        else:\n",
    "            ints[r] = [(h[\"start\"], h[\"end\"]) for h in res[r]]\n",
    "        \n",
    "    return ints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "35fd79e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "def readFasta(path):\n",
    "    return [r for r in SeqIO.parse(open(path, 'r'), \"fasta\")]\n",
    "\n",
    "#This function calculates the hash of a bitwise k-mer representation. The function is influenced by the code of \"The\n",
    "#minimizer Jaccard estimator is biased and inconsistent.\" from Belbasi et al. (function \"minimap2_hash(seed,v,mask)\"\n",
    "#in file \"minimap2_hash_uncompiled.py\").\n",
    "def getHash(kmer, mask):\n",
    "    u = kmer & mask\n",
    "    u = ((~u) + (u << 21)) & mask # u = (u<<21)-(u+1) = 77594587*u-1\n",
    "    u = u ^ (u >> 24)\n",
    "    u = ((u + (u << 3)) + (u << 8)) & mask # u *= 265\n",
    "    u = u ^ (u >> 14)\n",
    "    u = ((u + (u << 2)) + (u << 4)) & mask # u *= 21\n",
    "    u = u ^ (u >> 28)\n",
    "    u = (u + (u << 31)) & mask # u *= 2147483649\n",
    "\n",
    "    return u"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "39b78012",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This function returns a dictionary of k-mers from the given reference sketch that are covered by the given results \n",
    "#of some tool per each read from the given set of read ids.\n",
    "#refSk: A list of start positions of k-mers in a sequence that are chosen as being part of the sketch\n",
    "#res: A dictionary with read ids as keys and lists of start and stop coordinate tuples as values\n",
    "def getCovKmers(refSk, res):\n",
    "    covKmers = {}\n",
    "    #Sort start positions in sketch\n",
    "    sortedRefSk = sorted(refSk)\n",
    "    \n",
    "    #Iterate over reads:\n",
    "    for r in res:\n",
    "        #Initialize dictionary for this read\n",
    "        covKmers[r] = {}\n",
    "        \n",
    "        #If there are no results we are done for this read\n",
    "        if len(res[r]) == 0:\n",
    "            continue\n",
    "            \n",
    "        #Initialize a flag marking that we have reached the end of our result list\n",
    "        endReached = False\n",
    "        #Sort result coordinates for this read increasingly by start position\n",
    "        coords = sorted(res[r], key=lambda e: e[0])\n",
    "        #Initialize an index to know which result positions we consider currently (we start with the one that has \n",
    "        #the smallest start coordinate)\n",
    "        i = 0\n",
    "        \n",
    "        #Iterate over sorted list of start positions\n",
    "        for p in sortedRefSk:\n",
    "            #We need to switch to a result which has a larger start position as long as the position of the refer-\n",
    "            #ence sketch k-mer is larger than the end position of our current result\n",
    "            while p > coords[i][1]:\n",
    "                #Move to the next result\n",
    "                i += 1\n",
    "                \n",
    "                #Check if there is another result at all\n",
    "                if i >= len(coords):\n",
    "                    endReached = True\n",
    "                    break\n",
    "                    \n",
    "            #Check if we set the flag\n",
    "            if endReached:\n",
    "                break\n",
    "                \n",
    "            #Check if the k-mer starts before the current result\n",
    "            if p < coords[i][0]:\n",
    "                continue\n",
    "                \n",
    "            #If we have reached this point the k-mer must fall inside the result\n",
    "            covKmers[r][p] = None\n",
    "        \n",
    "    return covKmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "12e5e6a0",
   "metadata": {},
   "outputs": [],
   "source": [
    "K = 15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "06662712",
   "metadata": {},
   "outputs": [],
   "source": [
    "refSeq = readFasta(\"../simulations/genomes/t2thumanChrY.fasta\")[0].seq\n",
    "blKmers = {int(l): None for l in open(\"highAbundKmersMiniK15w10Lrgr100BtStrnds.txt\", 'r')}\n",
    "fltRfSkMiK15W10 = [k for k in calcMiniSketch(str(refSeq), K, 10) if k[2] not in blKmers]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d7a1cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load read sequences\n",
    "p = \"../simulations/reads/t2thumanChrY_sr0.00010909090909090909_dr0.0009818181818181818_i0.0009090909090909091_\" + \\\n",
    "\"sd7361077429744071834_lmn100_lmx1000000_lavg9000_ls7000_dp10.fasta\"\n",
    "rdSeqs = {r.id: r for r in readFasta(p)}\n",
    "fltRfSkMiKmrStPosK15W10 = [k[0] + 1 - K for k in fltRfSkMiK15W10 if k[2] not in blKmers]\n",
    "#Load homology intervals\n",
    "FindThomsNestResApprxMppng = {}\n",
    "bestFitHomInts = {}\n",
    "covKmersBestFitInts = {}\n",
    "p = \"../simulations/homologies/homologies_t2thumanChrY_sr0.000_dr0.001_i\" + \\\n",
    "\"0.001_sd7361077429744071834_lmn100_lmx1000000_lavg9000_ls7000_dp10_rm20_k15_w10_c1_u1_de\" + \\\n",
    "\"0.088_in-231.159_blhighAbundKmersMiniK15w10Lrgr100BtStrnds_nested.txt\"\n",
    "FindThomsNestResApprxMppng[20] = loadFindThomsRes(p, fltRfSkMiK15W10)\n",
    "bestFitHomInts[20] = filterBestFittingInts(parseCoordsFromFindThomsRes(FindThomsNestResApprxMppng[20], True))\n",
    "covKmersBestFitInts[20] = getCovKmers(fltRfSkMiKmrStPosK15W10, bestFitHomInts[20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd96ec32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
