{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3/dist-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.25.2\n",
      "  warnings.warn(f\"A NumPy version >={np_minversion} and <{np_maxversion}\"\n"
     ]
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pathlib import Path\n",
    "from IPython.display import display, HTML\n",
    "from Bio import SeqIO\n",
    "import os\n",
    "#from astropy import units as u\n",
    "import sys\n",
    "from readpaf import parse_paf\n",
    "from collections import Counter, defaultdict\n",
    "from tqdm import tqdm\n",
    "\n",
    "import sv\n",
    "\n",
    "#from itables import init_notebook_mode\n",
    "#init_notebook_mode(all_interactive=True)\n",
    "\n",
    "def to_latex(df, data, refname):\n",
    "    latex = \"\"\n",
    "    df.index = df.index.map(lambda x: f'\\\\{x}')\n",
    "    df.columns = df.columns.str.replace(' ', '\\\\\\\\')\n",
    "    df.columns = df.columns.str.replace('%', '\\%')\n",
    "    df.columns = df.columns.map(lambda x: '\\makecell{' + x + '}')\n",
    "    #df = df.astype(str).map(lambda x: x.rstrip('0').rstrip('.') if '.' in x else x)\n",
    "    latex += df.to_latex(escape=False, label=f'tab:{refname}', caption=data, float_format = lambda x: '{:0.2f}'.format(x) if pd.notna(x) else '-')\n",
    "    #latex += df.to_latex(float_format = lambda x: '{:0.2f}'.format(x) if pd.notna(x) else '-')\n",
    "    latex += '\\n'\n",
    "    return latex"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def union_length(df):\n",
    "    df_sorted = df.sort_values(by='target_start').reset_index(drop=True)\n",
    "    merged_intervals = []\n",
    "    current_start, current_end = df_sorted.iloc[0]['target_start'], df_sorted.iloc[0]['target_end']\n",
    "\n",
    "    for index, row in df_sorted.iterrows():\n",
    "        if index == 0:\n",
    "            continue\n",
    "        if row['target_start'] <= current_end:\n",
    "            current_end = max(current_end, row['target_end'])\n",
    "        else:\n",
    "            merged_intervals.append((current_start, current_end))\n",
    "            current_start, current_end = row['target_start'], row['target_end']\n",
    "\n",
    "    merged_intervals.append((current_start, current_end))\n",
    "    union_length = sum(end - start for start, end in merged_intervals)\n",
    "    return union_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def perc(a, b):\n",
    "    if b == 0:\n",
    "        return np.nan\n",
    "    return 100.0 * a / b\n",
    "\n",
    "def fasta2df(fn):\n",
    "    seqs = SeqIO.parse(fn, \"fasta\")\n",
    "    df = pd.DataFrame((str(s.id), str(s.seq)) for s in seqs)\n",
    "    df.columns = [\"ID\", \"Sequence\"]\n",
    "    return df\n",
    "\n",
    "#def is_overlapping(a, sv_row):\n",
    "#    return a.GT_from <= sv_row['END'] and sv_row['POS'] <= a.GT_to \n",
    "    \n",
    "min_overlap = 0.9\n",
    "\n",
    "def is_mapping_correct(a):\n",
    "     if a.GT_ref != a.target_name:\n",
    "         return False\n",
    "     if a.GT_strand != a.strand:\n",
    "         return False\n",
    "     union_from = min(a.GT_from, a.target_start)\n",
    "     union_to = max(a.GT_to, a.target_end)\n",
    "\n",
    "     intersect_from = max(a.GT_from, a.target_start)\n",
    "     intersect_to = min(a.GT_to, a.target_end)\n",
    "     overlaps = intersect_to - intersect_from >= min_overlap * (union_to - union_from)\n",
    "     return overlaps\n",
    "\n",
    "def read_paf(pref, reads, experiment, tool):\n",
    "    #display(vcf_df)\n",
    "    paf_file = pref.with_suffix('.paf')\n",
    "    no_GT = False\n",
    "    if not paf_file.exists():\n",
    "        raise Exception(f\"File does not exist or is empty: {paf_file}\")\n",
    "    with open(paf_file) as handle:\n",
    "        df = parse_paf(handle, dataframe=True)\n",
    "        df['experiment'] = experiment\n",
    "        df['tool'] = tool\n",
    "        try:\n",
    "            df[ ['read_name', 'GT_ref', 'GT_from', 'GT_to', 'GT_strand'] ] = df['query_name'].str.split('!', expand=True)\n",
    "            df['GT_from'] = df['GT_from'].astype(int)\n",
    "            df['GT_to'] = df['GT_to'].astype(int)\n",
    "            #df['is_correct_labels'] = df.apply(lambda x: is_correct_labels(x, orig_l, mutated_l), axis=1)\n",
    "            df['is_correct'] = df.apply(is_mapping_correct, axis=1)\n",
    "            #df['is_correct'] = df.apply(lambda x: is_correct_labels(x, orig_l, mutated_l), axis=1)\n",
    "            df['start_diff'] = df.target_start - df.GT_from  # TODO: different coordinate systems!\n",
    "            df['end_diff'] = df.target_end - df.GT_to  # TODO: different coordinate systems!\n",
    "            #df['read_sv'] = 'none' # df.apply(lambda x: read_falls_on_what_sv(x, vcf_df), axis=1)\n",
    "        except Exception as e:\n",
    "            display(e)\n",
    "            df['read_name'] = df['query_name']\n",
    "            #df['is_correct_labels'] = True\n",
    "            df['is_correct'] = True\n",
    "            df['start_diff'] = 0\n",
    "            df['end_diff'] = 0\n",
    "            #df['read_sv'] = 'none'\n",
    "            no_GT = True\n",
    "    df = df.sort_values(['read_name', 'residue_matches'], ascending=[True, False], ignore_index=True)\n",
    "    display(df)\n",
    "\n",
    "    # initialize from left to right\n",
    "    myNA = 'N/A'\n",
    "    #bad = 'bad'\n",
    "    unique = 'uniq'\n",
    "    multi = 'mult'\n",
    "    #bad_mappings_perc = bad+' mappings p.multiread'\n",
    "    paf = defaultdict(int)\n",
    "    #paf['unmapped'] = 0\n",
    "    #paf[unique] = 0\n",
    "    #paf[multi] = 0\n",
    "    #for suffix in ['', ' Q60']:\n",
    "    #    paf[bad + suffix] = 0\n",
    "    #    #if suffix == '':\n",
    "    #        #paf[multi+' '+bad+' group'] = 0\n",
    "    #        #paf[bad_mappings_perc] = 0\n",
    "    #    for unique_or_multi in [unique, multi]:\n",
    "    #        paf[bad + ' ' + unique_or_multi + suffix] = 0\n",
    "\n",
    "    def bad(sv_type):\n",
    "        return 'bad ' + sv_type\n",
    "\n",
    "    def process_group(group_first_index, group_last_index):\n",
    "        group = df.loc[group_first_index:group_last_index]\n",
    "        first_in_group = df.loc[group_first_index]\n",
    "        group_is_correct = group['is_correct'].any()\n",
    "        group_is_unique = len(group) == 1\n",
    "        suffix = 'Q60' if (group.mapping_quality == 60).all() else ''\n",
    "\n",
    "        unique_or_multi = unique if group_is_unique else multi\n",
    "        paf[unique_or_multi] += 1\n",
    "        if not group_is_correct:\n",
    "            paf[bad(suffix)] += 1\n",
    "            paf[unique_or_multi + ' ' +bad(suffix)] += 1\n",
    "            #if not group_is_unique:\n",
    "            #    paf[multi+' '+bad+' group'] += 1\n",
    "            #    paf[bad_mappings_perc] += 1\n",
    "            #paf[bad(first_in_group['read_sv'])] += 1\n",
    "        #paf[first_in_group['read_sv']] += 1\n",
    "\n",
    "    group_first_i, group_read_name = 0, df.loc[0, 'read_name']\n",
    "    for i, a in df.iterrows():\n",
    "        if a.read_name != group_read_name:\n",
    "            process_group(group_first_i, i-1)\n",
    "            group_first_i, group_read_name = i, a.read_name\n",
    "    process_group(group_first_i, len(df)-1)\n",
    "\n",
    "    paf['unmapped'] = reads - paf[unique] - paf[multi]\n",
    "\n",
    "    paf['aligned_with 0/1/2+_segments'] = '{} / {} / {} ({:.0f}%/{:.0f}%/{:.0f}%)'.format(\n",
    "        paf['unmapped'], paf[unique], paf[multi],\n",
    "        perc(paf['unmapped'], len(df)), perc(paf[unique], len(df)), perc(paf[multi], len(df)))\n",
    "    #for key in [unique, multi, 'unmapped', bad]:\n",
    "    #    paf[key] = '{} ({:.2f}%)'.format(paf[key], perc(paf[key], reads))\n",
    "\n",
    "    #if paf[multi] > 0:\n",
    "    #    paf[bad_mappings_perc] = '{:.2f}'.format(paf[bad_mappings_perc] / paf[multi])\n",
    "    #else:\n",
    "    #    paf[bad_mappings_perc] = myNA\n",
    "\n",
    "    paf['mean start diff'] = '{:.1f}'.format(df[df.is_correct].start_diff.mean())\n",
    "    paf['mean end diff'] = '{:.1f}'.format(df[df.is_correct].end_diff.mean())\n",
    " \n",
    "    #if no_GT:\n",
    "    #    for suff in ['', ' '+unique, ' '+multi]:\n",
    "    #        for suffsuff in ['', ' Q60']:\n",
    "    #            paf[bad + suff + suffsuff] = myNA \n",
    "    #        #paf[multi+' '+bad+' group'] = myNA\n",
    "    #        paf['addit '+bad+' aligns'] = myNA\n",
    "    #    paf['mean start diff'] = myNA\n",
    "    #    paf['mean end diff'] = myNA\n",
    "\n",
    "    paf['alignments depth'] = '{:.2f}'.format(sum(df['query_length']) / df.iloc[0]['target_length'])\n",
    "    union_len = union_length(df)\n",
    "    paf['covered reference'] = '{} ({:.2f}%)'.format(union_len, perc(union_len, df.iloc[0]['target_length']))\n",
    "\n",
    "    paf.pop('unmapped')\n",
    "    paf.pop(unique)\n",
    "    paf.pop(multi)\n",
    "\n",
    "    return pd.Series(paf, dtype='object')\n",
    "    \n",
    "def read_times(pref):\n",
    "    times = {}\n",
    "    with open(str(pref) + '.index.time') as f_index_time:\n",
    "        index_time, index_mem = map(float, f_index_time.readline().split())\n",
    "        times['index time'] = index_time #* u.second\n",
    "        #times['index_mem'] = index_mem / 2**20\n",
    "        with open(str(pref) + '.time') as f_time:\n",
    "            total_time, total_mem = map(float, f_time.readline().split())\n",
    "            #times['time total'] = total_time #* u.second\n",
    "            times['map time'] = total_time - times['index time']\n",
    "            times['memory'] = (total_mem / 2**20) #* u.GB\n",
    "    return pd.Series(times, dtype='object').map('{:.1f}'.format)\n",
    "\n",
    "def get_comparison_table(main_dir: Path, refname, experiment: Path, tools):\n",
    "    empty_cell = -1\n",
    "    alldf = pd.DataFrame()\n",
    "    alldf.name = experiment\n",
    "    ref = fasta2df(Path('refs') / (refname + '.fa'))\n",
    "    reads = fasta2df(Path('reads') / (str(experiment) + '.fa'))\n",
    "\n",
    "    rows = []\n",
    "\n",
    "    for tool in tqdm(tools, desc=f'Tools for {experiment}', leave=False):\n",
    "        d = Path(main_dir) / experiment / tool / tool\n",
    "        row = pd.Series({\n",
    "            'tool': tool,\n",
    "            'reads': len(reads),\n",
    "        })\n",
    "        row = pd.concat([row, read_paf(d, len(reads), experiment, tool)])     # .paf\n",
    "        try:\n",
    "            row = pd.concat([row, read_times(d)])   # .time, .index.time\n",
    "        except Exception as e:\n",
    "            print(f\"An error occurred while reading times {d}: {e}\")\n",
    "            row['index time'] = empty_cell\n",
    "            row['map time'] = empty_cell\n",
    "            row['memory'] = empty_cell\n",
    "        rows.append(row)\n",
    "    alldf = pd.DataFrame(rows)\n",
    "    alldf = alldf.set_index('tool')\n",
    "    alldf.index.name = None\n",
    "    return alldf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Build a table to compare the mappers by accuracy, runtime (indexing and mapping) and memory."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       " <style> table { font-family: \"Courier New\", Courier, monospace; } </style> "
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tools for t2tChrY-readst2tChrY-a0.99-d10-l10000:   0%|          | 0/1 [00:00<?, ?it/s]"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>query_name</th>\n",
       "      <th>query_length</th>\n",
       "      <th>query_start</th>\n",
       "      <th>query_end</th>\n",
       "      <th>strand</th>\n",
       "      <th>target_name</th>\n",
       "      <th>target_length</th>\n",
       "      <th>target_start</th>\n",
       "      <th>target_end</th>\n",
       "      <th>residue_matches</th>\n",
       "      <th>...</th>\n",
       "      <th>experiment</th>\n",
       "      <th>tool</th>\n",
       "      <th>read_name</th>\n",
       "      <th>GT_ref</th>\n",
       "      <th>GT_from</th>\n",
       "      <th>GT_to</th>\n",
       "      <th>GT_strand</th>\n",
       "      <th>is_correct</th>\n",
       "      <th>start_diff</th>\n",
       "      <th>end_diff</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S1_1!NC_060948.1!41948194!41955717!+</td>\n",
       "      <td>7542</td>\n",
       "      <td>0</td>\n",
       "      <td>7541</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>47189864</td>\n",
       "      <td>47198068</td>\n",
       "      <td>7542</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_1</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>41948194</td>\n",
       "      <td>41955717</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>5241670</td>\n",
       "      <td>5242351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S1_10!NC_060948.1!21496431!21502430!-</td>\n",
       "      <td>6014</td>\n",
       "      <td>0</td>\n",
       "      <td>6013</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>21496478</td>\n",
       "      <td>21502363</td>\n",
       "      <td>6014</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_10</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>21496431</td>\n",
       "      <td>21502430</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>47</td>\n",
       "      <td>-67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S1_100!NC_060948.1!31615896!31625898!-</td>\n",
       "      <td>10021</td>\n",
       "      <td>0</td>\n",
       "      <td>10020</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>31614936</td>\n",
       "      <td>31625408</td>\n",
       "      <td>10021</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_100</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>31615896</td>\n",
       "      <td>31625898</td>\n",
       "      <td>-</td>\n",
       "      <td>False</td>\n",
       "      <td>-960</td>\n",
       "      <td>-490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S1_1000!NC_060948.1!21434231!21444413!-</td>\n",
       "      <td>10211</td>\n",
       "      <td>0</td>\n",
       "      <td>10210</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>21434263</td>\n",
       "      <td>21444042</td>\n",
       "      <td>10211</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_1000</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>21434231</td>\n",
       "      <td>21444413</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>32</td>\n",
       "      <td>-371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S1_10000!NC_060948.1!59453112!59462344!-</td>\n",
       "      <td>9264</td>\n",
       "      <td>0</td>\n",
       "      <td>9263</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>59452754</td>\n",
       "      <td>59462389</td>\n",
       "      <td>9264</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_10000</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>59453112</td>\n",
       "      <td>59462344</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>-358</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62460</th>\n",
       "      <td>S1_9995!NC_060948.1!48317964!48328285!+</td>\n",
       "      <td>10341</td>\n",
       "      <td>0</td>\n",
       "      <td>10340</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>48317012</td>\n",
       "      <td>48328243</td>\n",
       "      <td>10341</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_9995</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>48317964</td>\n",
       "      <td>48328285</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>-952</td>\n",
       "      <td>-42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62461</th>\n",
       "      <td>S1_9996!NC_060948.1!9696138!9705900!-</td>\n",
       "      <td>9805</td>\n",
       "      <td>0</td>\n",
       "      <td>9804</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>9696161</td>\n",
       "      <td>9705883</td>\n",
       "      <td>9805</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_9996</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>9696138</td>\n",
       "      <td>9705900</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>-17</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62462</th>\n",
       "      <td>S1_9997!NC_060948.1!29720724!29731033!+</td>\n",
       "      <td>10332</td>\n",
       "      <td>0</td>\n",
       "      <td>10331</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>29719716</td>\n",
       "      <td>29730605</td>\n",
       "      <td>10332</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_9997</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>29720724</td>\n",
       "      <td>29731033</td>\n",
       "      <td>+</td>\n",
       "      <td>False</td>\n",
       "      <td>-1008</td>\n",
       "      <td>-428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62463</th>\n",
       "      <td>S1_9998!NC_060948.1!61846335!61852559!-</td>\n",
       "      <td>6248</td>\n",
       "      <td>0</td>\n",
       "      <td>6247</td>\n",
       "      <td>-</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>61846048</td>\n",
       "      <td>61852592</td>\n",
       "      <td>6248</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_9998</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>61846335</td>\n",
       "      <td>61852559</td>\n",
       "      <td>-</td>\n",
       "      <td>True</td>\n",
       "      <td>-287</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62464</th>\n",
       "      <td>S1_9999!NC_060948.1!20581596!20592447!+</td>\n",
       "      <td>10891</td>\n",
       "      <td>0</td>\n",
       "      <td>10890</td>\n",
       "      <td>+</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>62460029</td>\n",
       "      <td>20581619</td>\n",
       "      <td>20592393</td>\n",
       "      <td>10891</td>\n",
       "      <td>...</td>\n",
       "      <td>t2tChrY-readst2tChrY-a0.99-d10-l10000</td>\n",
       "      <td>jaccmap</td>\n",
       "      <td>S1_9999</td>\n",
       "      <td>NC_060948.1</td>\n",
       "      <td>20581596</td>\n",
       "      <td>20592447</td>\n",
       "      <td>+</td>\n",
       "      <td>True</td>\n",
       "      <td>23</td>\n",
       "      <td>-54</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>62465 rows × 42 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     query_name  query_length  query_start  query_end strand  \\\n",
       "0          S1_1!NC_060948.1!41948194!41955717!+          7542            0       7541      +   \n",
       "1         S1_10!NC_060948.1!21496431!21502430!-          6014            0       6013      -   \n",
       "2        S1_100!NC_060948.1!31615896!31625898!-         10021            0      10020      -   \n",
       "3       S1_1000!NC_060948.1!21434231!21444413!-         10211            0      10210      -   \n",
       "4      S1_10000!NC_060948.1!59453112!59462344!-          9264            0       9263      -   \n",
       "...                                         ...           ...          ...        ...    ...   \n",
       "62460   S1_9995!NC_060948.1!48317964!48328285!+         10341            0      10340      +   \n",
       "62461     S1_9996!NC_060948.1!9696138!9705900!-          9805            0       9804      -   \n",
       "62462   S1_9997!NC_060948.1!29720724!29731033!+         10332            0      10331      +   \n",
       "62463   S1_9998!NC_060948.1!61846335!61852559!-          6248            0       6247      -   \n",
       "62464   S1_9999!NC_060948.1!20581596!20592447!+         10891            0      10890      +   \n",
       "\n",
       "       target_name  target_length  target_start  target_end  residue_matches  ...  \\\n",
       "0      NC_060948.1       62460029      47189864    47198068             7542  ...   \n",
       "1      NC_060948.1       62460029      21496478    21502363             6014  ...   \n",
       "2      NC_060948.1       62460029      31614936    31625408            10021  ...   \n",
       "3      NC_060948.1       62460029      21434263    21444042            10211  ...   \n",
       "4      NC_060948.1       62460029      59452754    59462389             9264  ...   \n",
       "...            ...            ...           ...         ...              ...  ...   \n",
       "62460  NC_060948.1       62460029      48317012    48328243            10341  ...   \n",
       "62461  NC_060948.1       62460029       9696161     9705883             9805  ...   \n",
       "62462  NC_060948.1       62460029      29719716    29730605            10332  ...   \n",
       "62463  NC_060948.1       62460029      61846048    61852592             6248  ...   \n",
       "62464  NC_060948.1       62460029      20581619    20592393            10891  ...   \n",
       "\n",
       "                                  experiment     tool  read_name       GT_ref   GT_from     GT_to  \\\n",
       "0      t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap       S1_1  NC_060948.1  41948194  41955717   \n",
       "1      t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap      S1_10  NC_060948.1  21496431  21502430   \n",
       "2      t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap     S1_100  NC_060948.1  31615896  31625898   \n",
       "3      t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap    S1_1000  NC_060948.1  21434231  21444413   \n",
       "4      t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap   S1_10000  NC_060948.1  59453112  59462344   \n",
       "...                                      ...      ...        ...          ...       ...       ...   \n",
       "62460  t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap    S1_9995  NC_060948.1  48317964  48328285   \n",
       "62461  t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap    S1_9996  NC_060948.1   9696138   9705900   \n",
       "62462  t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap    S1_9997  NC_060948.1  29720724  29731033   \n",
       "62463  t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap    S1_9998  NC_060948.1  61846335  61852559   \n",
       "62464  t2tChrY-readst2tChrY-a0.99-d10-l10000  jaccmap    S1_9999  NC_060948.1  20581596  20592447   \n",
       "\n",
       "       GT_strand  is_correct  start_diff  end_diff  \n",
       "0              +       False     5241670   5242351  \n",
       "1              -        True          47       -67  \n",
       "2              -       False        -960      -490  \n",
       "3              -        True          32      -371  \n",
       "4              -        True        -358        45  \n",
       "...          ...         ...         ...       ...  \n",
       "62460          +        True        -952       -42  \n",
       "62461          -        True          23       -17  \n",
       "62462          +       False       -1008      -428  \n",
       "62463          -        True        -287        33  \n",
       "62464          +        True          23       -54  \n",
       "\n",
       "[62465 rows x 42 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "                                                                                              \r"
     ]
    },
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "</style>\n",
       "<table id=\"T_7829b\">\n",
       "  <caption>t2tChrY-readst2tChrY-a0.99-d10-l10000</caption>\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_7829b_level0_col0\" class=\"col_heading level0 col0\" >reads</th>\n",
       "      <th id=\"T_7829b_level0_col1\" class=\"col_heading level0 col1\" >bad </th>\n",
       "      <th id=\"T_7829b_level0_col2\" class=\"col_heading level0 col2\" >uniq bad </th>\n",
       "      <th id=\"T_7829b_level0_col3\" class=\"col_heading level0 col3\" >bad Q60</th>\n",
       "      <th id=\"T_7829b_level0_col4\" class=\"col_heading level0 col4\" >uniq bad Q60</th>\n",
       "      <th id=\"T_7829b_level0_col5\" class=\"col_heading level0 col5\" >aligned_with 0/1/2+_segments</th>\n",
       "      <th id=\"T_7829b_level0_col6\" class=\"col_heading level0 col6\" >mean start diff</th>\n",
       "      <th id=\"T_7829b_level0_col7\" class=\"col_heading level0 col7\" >mean end diff</th>\n",
       "      <th id=\"T_7829b_level0_col8\" class=\"col_heading level0 col8\" >alignments depth</th>\n",
       "      <th id=\"T_7829b_level0_col9\" class=\"col_heading level0 col9\" >covered reference</th>\n",
       "      <th id=\"T_7829b_level0_col10\" class=\"col_heading level0 col10\" >index time</th>\n",
       "      <th id=\"T_7829b_level0_col11\" class=\"col_heading level0 col11\" >map time</th>\n",
       "      <th id=\"T_7829b_level0_col12\" class=\"col_heading level0 col12\" >memory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_7829b_level0_row0\" class=\"row_heading level0 row0\" >jaccmap</th>\n",
       "      <td id=\"T_7829b_row0_col0\" class=\"data row0 col0\" >62465</td>\n",
       "      <td id=\"T_7829b_row0_col1\" class=\"data row0 col1\" >5567</td>\n",
       "      <td id=\"T_7829b_row0_col2\" class=\"data row0 col2\" >5567</td>\n",
       "      <td id=\"T_7829b_row0_col3\" class=\"data row0 col3\" >2875</td>\n",
       "      <td id=\"T_7829b_row0_col4\" class=\"data row0 col4\" >2875</td>\n",
       "      <td id=\"T_7829b_row0_col5\" class=\"data row0 col5\" >0 / 62465 / 0 (0%/100%/0%)</td>\n",
       "      <td id=\"T_7829b_row0_col6\" class=\"data row0 col6\" >-247.4</td>\n",
       "      <td id=\"T_7829b_row0_col7\" class=\"data row0 col7\" >-25.5</td>\n",
       "      <td id=\"T_7829b_row0_col8\" class=\"data row0 col8\" >10.00</td>\n",
       "      <td id=\"T_7829b_row0_col9\" class=\"data row0 col9\" >62391540 (99.89%)</td>\n",
       "      <td id=\"T_7829b_row0_col10\" class=\"data row0 col10\" >0.7</td>\n",
       "      <td id=\"T_7829b_row0_col11\" class=\"data row0 col11\" >117.1</td>\n",
       "      <td id=\"T_7829b_row0_col12\" class=\"data row0 col12\" >0.4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x76791bdeca90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "main_dir = Path('out20241022')\n",
    "#tools = ['sweepmap', 'sweepmap-slow', 'mapquik', 'blend', 'minimap', 'winnowmap', 'rmqmap'] \n",
    "#tools = ['minimap', 'mapquik', 'blend', 'jaccmap']\n",
    "tools = ['jaccmap']\n",
    "experiments = [\n",
    "    ('t2tChrY', 't2tChrY-readst2tChrY-a0.99-d10-l10000'),\n",
    "#    ('chm13',   'chm13-readschm13-a0.99-d0.1-l10000'),\n",
    "#    ('t2tChrY', 't2tChrY-readst2tChrY-a0.99-d1-l24000'),\n",
    "#    ('chm13',   'HG002_24kb'),\n",
    "]\n",
    "\n",
    "pd.set_option('display.width', 100)\n",
    "css = \"\"\" <style> table { font-family: \"Courier New\", Courier, monospace; } </style> \"\"\"\n",
    "display(HTML(css))\n",
    "dfs = []\n",
    "keys = []\n",
    "for refname, data in experiments:\n",
    "    df = get_comparison_table(main_dir=main_dir, refname=refname, experiment=data, tools=tools).round(2)\n",
    "    dfs.append(df)\n",
    "    keys.append(data)\n",
    "    df_styled = df.style.set_caption(data)\n",
    "    display(df_styled)\n",
    "    #print(vcf_df.to_latex(escape=True))\n",
    "#DF = pd.concat(dfs, keys=keys)\n",
    "#display(DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#def is_correct_labels(a, GT_l, target_l, debug=False):\n",
    "#    if a.GT_ref != a.target_name:\n",
    "#        return False\n",
    "#    if a.GT_strand != a.strand:  # won't work for inversions\n",
    "#        return False\n",
    "##    union_from = min(a.GT_from, a.target_start)\n",
    "##    union_to = max(a.GT_to, a.target_end)\n",
    "#\n",
    "#    GT_labels = Counter(target_l[a.GT_ref][a.GT_from:a.GT_to])\n",
    "#    target_labels = Counter(GT_l[a.target_name][a.target_start:a.target_end])\n",
    "#    #GT_labels = Counter(GT_l[a.GT_ref][a.GT_from:a.GT_to])\n",
    "#    #target_labels = Counter(target_l[a.target_name][a.target_start:a.target_end])\n",
    "#    intersection = GT_labels & target_labels\n",
    "#    union = GT_labels | target_labels\n",
    "#    overlaps = sum(intersection.values()) >= min_overlap * sum(union.values())\n",
    "#\n",
    "#    if debug:\n",
    "#        display(a)\n",
    "#        display('           GT from {} to {}'.format(a.GT_from, a.GT_to))\n",
    "#        display('       target from {} to {}'.format(a.target_start, a.target_end))\n",
    "#        display('    GT_labels from {} to {}'.format(min(GT_labels), max(GT_labels)))\n",
    "#        display('target_labels from {} to {}'.format(min(target_labels), max(target_labels)))\n",
    "#        display('{} >?= {} = {} * union {} => {}'.format(sum(intersection.values()), min_overlap * sum(union.values()), min_overlap, sum(union.values()), overlaps))\n",
    "#    return overlaps\n",
    "\n",
    "#def is_correct_labels_df(df: pd.DataFrame, GT_l, target_l, debug=False):\n",
    "#    if (df.GT_ref != df.target_name).any():\n",
    "#        return False\n",
    "#    if (df.GT_strand != df.strand).any():  # won't work for inversions\n",
    "#        return False\n",
    "#\n",
    "#    GT_ref, GT_from, GT_to = df.GT_ref.iloc[0], df.GT_from.iloc[0], df.GT_to.iloc[0]\n",
    "#    assert (df.GT_ref == GT_ref).all() and (df.GT_from == GT_from).all() and (df.GT_to == GT_to).all()\n",
    "#\n",
    "#    GT_labels = Counter(target_l[GT_ref][GT_from:GT_to])\n",
    "#    L = [ Counter(GT_l[a.target_name][a.target_start:a.target_end]) for a in df.itertuples() ]\n",
    "#    target_labels = sum(L, Counter())\n",
    "#    #GT_labels = Counter(GT_l[a.GT_ref][a.GT_from:a.GT_to])\n",
    "#    #target_labels = Counter(target_l[a.target_name][a.target_start:a.target_end])\n",
    "#    intersection = GT_labels & target_labels\n",
    "#    union = GT_labels | target_labels\n",
    "#    overlaps = sum(intersection.values()) >= min_overlap * sum(union.values())\n",
    "#    return overlaps\n",
    "\n",
    "#def read_falls_on_what_sv(a, vcf_df):\n",
    "#    query_start = a['GT_from']\n",
    "#    query_end = a['GT_to']\n",
    "#    start_idx = vcf_df['POS'].searchsorted(query_start, side='right')\n",
    "#    end_idx = vcf_df['END'].searchsorted(query_end, side='left')\n",
    "#    overlap_df = vcf_df.iloc[start_idx:end_idx]\n",
    "#    overlap_ch_df = overlap_df[overlap_df['CHROM'] == a['GT_ref']]\n",
    "#    overlap_ch_sv = overlap_ch_df[(overlap_ch_df['POS'] <= query_end) & (overlap_ch_df['END'] >= query_start)]\n",
    "#    #if len(overlap_ch_sv) > 0:\n",
    "#    #    display(a, overlap_ch_sv)\n",
    "#    if len(overlap_ch_sv) == 0:\n",
    "#        return 'none'\n",
    "#    elif len(overlap_ch_sv) == 1:\n",
    "#        return overlap_ch_sv.iloc[0]['SVTYPE']\n",
    "#    else:\n",
    "#        return 'multi'\n",
    "\n",
    "    # SVs\n",
    "    #orig_fa_dict = sv.read_fasta_file(Path('refs') / (refname+'.fa'))\n",
    "    #orig_l = sv.gen_unique_labels(orig_fa_dict)\n",
    "    #try:\n",
    "    #    vcf_df = sv.read_vcf(Path('refs') / (readsim_refname+'.vcf'))\n",
    "    #    assert((vcf_df.POS <= vcf_df.END).all())\n",
    "    #    vcf_df = vcf_df.sort_values(by='POS').reset_index(drop=True)\n",
    "    #    mutated_fa, mutated_l = sv.mutate(orig_fa_dict, orig_l, vcf_df)\n",
    "    #except FileNotFoundError:\n",
    "    #    vcf_df = sv.read_vcf(Path('refs') / 'empty.vcf')\n",
    "    #    mutated_fa, mutated_l = orig_fa_dict, orig_l\n",
    "\n",
    "\n",
    "    #print(vcf_df)\n",
    "    #for sv_type in list(vcf_df['SVTYPE'].unique()) + ['none', 'multi']:\n",
    "    #    paf[bad(sv_type)] = '{} / {} ({:.2f}%)'.format(paf[bad(sv_type)], paf[sv_type], perc(paf[bad(sv_type)], paf[sv_type]))\n",
    "    #    paf.pop(sv_type)\n",
    "\n",
    "    # debug\n",
    "    #for row in df[df['read_sv'] != 'none'].itertuples():\n",
    "    #    is_correct_labels(row, orig_l, mutated_l, debug=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
